src/change_stream/document.rs

```diff
@@ -0,0 +1,101 @@
+//! Contains documents related to a ChangeStream event.
+use crate::coll::Namespace;
+use bson::{Bson, Document};
+use serde::{Deserialize, Serialize};
+
+/// An opaque token used for resuming an interrupted `ChangeStream`.
+///
+/// When starting a new change stream, `start_after` and `resume_after` fields on
+/// `ChangeStreamOptions` can be specified with instances of `ResumeToken`.
+///
+/// See the documentation
+/// [here](https://docs.mongodb.com/manual/changeStreams/#change-stream-resume-token) for more
+/// information on resume tokens.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ResumeToken(pub(crate) Bson);
+
+/// A `ChangeStreamEventDocument` represents a
+/// [change event](https://docs.mongodb.com/manual/reference/change-events/) in the associated change stream.
+/// Instances of `ChangeStreamEventDocument` are returned from calls to `ChangeStream::next`.
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub struct ChangeStreamEventDocument {
+    /// An opaque token for use when resuming an interrupted `ChangeStream`.
+    ///
+    /// See the documentation
+    /// [here](https://docs.mongodb.com/manual/changeStreams/#change-stream-resume-token) for
+    /// more information on resume tokens.
+    ///
+    /// Also see the documentation on [resuming a change
+    /// stream](https://docs.mongodb.com/manual/changeStreams/#resume-a-change-stream).
+    #[serde(rename = "_id")]
+    pub id: ResumeToken,
+
+    /// Describes the type of operation represented in this change notification.
+    pub operation_type: OperationType,
+
+    /// Contains two fields: "db" and "coll" containing the database and collection name in which
+    /// the change happened.
+    pub ns: Option<Namespace>,
```

Comment:
This is missing the [`to` field](https://docs.mongodb.com/manual/reference/change-events/#rename-event) for rename events. The spec also needs to be updated to include it too, so I filed [DRIVERS-1343](https://jira.mongodb.org/browse/DRIVERS-1343) for that.

Reply:
Should `to` be `Option<Document>` or `Option<Namespace>`?

Reply:
The link suggests that to is of document type. so using that. fixed

Reply:
I tend to think `to` should be an `Option<Namespace>`, since that field contains the namespace a given collection is being renamed to. Ironically, I'm now realizing that I think the `ns` field can't be a `Namespace`, since the `coll` part will be missing in `dropDatabase` events. For consistency, we could consider introducing a new struct that both of them use with an optional `coll` portion.

Reply:
Sounds good

Reply:
Hm, what about actually just using `Namespace` for both and having `coll` have a serde default of empty string?

Reply:
added an ChangeStreamEventSource enum with a namespace and db. fixed